README.md

sbt-jetty-embed
===============

This is an sbt plugin to create 'executeable' war files by embedding jetty.
It supports both Jetty-6 & Jetty-7.

This plugin takes a brute force approach, it copies a startup source file
into a local managed source tree, which provides the jetty embed code.  It
then overrides the webapp prepare task to also unpack the jetty jars into
the root of the war file.

Important Note
--------------

Requires SBT 0.7.5RC0 or later.  There are bugs in earlier versions of sbt
which will impact the behaviour of this plugin.

Lift users: if you are using Lift 2.3-M1 and newer it automatically detects
Jetty 6, Jetty 7, and Servlet 3.0.  However earlier versions should be aware
that to use Jetty-7 you will need an addition to your Boot.scala file.
Please see the example lift-project.

Further Information
-------------------

See SSL.md for information regarding using SSL.
See DEVELOPER.md for information regarding building manually.


How to use in your own project
------------------------------

First create the plugins directory and file

	mkdir project/plugins
	vi project/plugins/Plugins.scala

	import sbt._

	class Plugins(info: ProjectInfo) extends PluginDefinition(info) {
		val jettyEmbeddedWarRepo = "Embeded Jetty Repo" at "https://github.com/glenford/repo/raw/master"
  		val jettyEmbeddedWar = "net.usersource" % "sbt-jetty-embed-plugin" % "0.6.1"
	}

Extend the plugin instead of DefaultWebProject

	import sbt._
	import net.usersource.jettyembed._
	
	class BasicProject(info :ProjectInfo) extends JettyEmbedWebProject(info)


By default it uses Jetty-6, if you would like to use Jetty-7 then use

	class BasicProject(info :ProjectInfo) extends JettyEmbedWebProject(info,JETTY7)



Update to ensure the jetty libs are downloaded and extract the startup code into your project

	sbt
	> update


This will copy a Startup file to

	src_managed/main/java/net/usersource/jettyembed/jetty6/Startup.java 


Then package your code as normal

	sbt
	> compile
	> package

Then run your new war

	$ java -jar target/scala_2.8.1/sbt-jetty-embed-basic_2.8.1-0.6.1.war 
	2011-03-03 21:10:56.157:INFO::Logging to STDERR via org.mortbay.log.StdErrLog
	2011-03-03 21:10:56.158:INFO::jetty-6.1.x
	2011-03-03 21:10:56.209:INFO::Extract file:/Users/glen/UserSource/Public/sbt-jetty-embed/basic-project/target/scala_2.8.1/sbt-jetty-embed-basic_2.8.1-0.6.1.war to /var/folders/Kg/KgM30VT2FvaHmZ1PuJ4D9++++TI/-Tmp-/Jetty_0_0_0_0_8080_sbt.jetty.embed.basic_2.8.1-0.6.1.war____.a3s7p1/webapp
	2011-03-03 21:10:57.569:INFO::NO JSP Support for /, did not find org.apache.jasper.servlet.JspServlet
	2011-03-03 21:10:57.754:INFO::Started SelectChannelConnector@0.0.0.0:8080


You can still use your 'executeable' war in a regular web container unchanged.


Runtime Options
---------------

You can override the default port of 8080 by using -DjettyPort e.g.

	$ java -DjettyPort=8000 -jar target/scala_2.8.1/sbt-jetty-embed-basic_2.8.1-0.6.1.war

You can set the temporary directory (where jetty unpacks the war) by using -DjettyTempDir e.g.

	$ java -DjettyTempDir=myTempDir -jar target/scala_2.8.1/sbt-jetty-embed-basic_2.8.1-0.6.1.war

You can override the default max idle time of 30000 milliseconds by using -DjettyMaxIdle e.g.

	$ java -DjettyMaxIdle=60000 -jar target/scala_2.8.1/sbt-jetty-embed-basic_2.8.1-0.6.1.war

You can run in interactive mode so you can press any key to exit by using -DjettyInteractive=true (default is false)

	$ java -DjettyInteractive=true -jar target/scala_2.8.1/sbt-jetty-embed-basic_2.8.1-0.6.1.war

By default Jetty will use NIO, you can turn this off by using

        $ java -DjettyNio=false -jar target/scala_2.8.1/sbt-jetty-embed-basic_2.8.1-0.6.1.war

You can enable debug logging, both for the startup and Jetty by using

	$ java -DjettyDebug=true -jar target/scala_2.8.1/sbt-jetty-embed-basic_2.8.1-0.6.1.war


Using a different version of Jetty
----------------------------------

You can override the default 6.1.22 version of Jetty to use by:

	override val jettyEmbedVersion = "6.1.26"

In your project definition.


License
-------

The sbt-jetty-embed plugin is licensed under the Apache 2 License
Please see the LICENSE and NOTICE files.




RELEASES.md

0.6.1
-----

* fix for SSL parameters

0.6
---

* fix for Issue-10
* added debug option

0.5
---

* fix for Issue-8 with multi-projects

0.4.1
-----

* moved startup source code to src_managed directory, rather than users src tree


0.4
---

* cleaned up naming to ensure it is named sbt-jetty-embed-plugin so not to potentially confuse with any other plugin type
* no longer need to use 'jetty-embed-prepare'
* no longer a SNAPSHOT build
* a couple of small refactorings


0.3-SNAPSHOT
------------

* added support for Jetty-7
* some small refactorings



Earlier versions were not published to the repository



plugin/src/main/java/net/usersource/jettyembed/jetty7/Startup.java

//
// Copyright 2011, Glen Ford
//
// Apache 2.0 License
// Please see README.md, LICENSE and NOTICE
//


package net.usersource.jettyembed.jetty7;

import org.eclipse.jetty.server.Connector;
import org.eclipse.jetty.server.Server;
import org.eclipse.jetty.server.bio.SocketConnector;
import org.eclipse.jetty.server.nio.SelectChannelConnector;
import org.eclipse.jetty.server.ssl.SslConnector;
import org.eclipse.jetty.server.ssl.SslSelectChannelConnector;
import org.eclipse.jetty.server.ssl.SslSocketConnector;
import org.eclipse.jetty.webapp.WebAppClassLoader;
import org.eclipse.jetty.webapp.WebAppContext;
import org.eclipse.jetty.util.URIUtil;

import java.io.File;
import java.io.IOException;
import java.net.URL;
import java.security.ProtectionDomain;




public class Startup {

    private static final String JETTY_PORT_NAME = "jettyPort";
    private static final String JETTY_SSL_PORT_NAME = "jettySslPort";
    private static final String JETTY_SSL_KEY_PASSWORD_NAME = "jettySslKeyPassword";
    private static final String JETTY_SSL_KEY_STOREFILE_NAME = "jettySslKeyStoreFile";
    private static final String JETTY_INTERACTIVE_NAME = "jettyInteractive";
    private static final String JETTY_USE_NIO_NAME = "jettyNio";
    private static final String JETTY_DEBUG_NAME = "jettyDebug";
    private static final String JETTY_MAX_IDLE_NAME = "jettyMaxIdle";
    
    private static boolean debug = false;
    private static boolean isInteractive = false;
    private static boolean usingNIO = true;
    private static boolean usingSSL = false;
    private static String tempDir = null;
    private static int jettyPort = 8080;
    private static int jettySSLPort = 0;
    private static String keyFileName = null;
    private static String keyPassword = null;
    private static int maxIdleTime = 30000;



    private static void processOptions() {

        if( System.getProperty(JETTY_DEBUG_NAME) != null ) debug = Boolean.getBoolean(JETTY_DEBUG_NAME);

        if( System.getProperty(JETTY_USE_NIO_NAME) != null ) usingNIO = Boolean.getBoolean(JETTY_USE_NIO_NAME);

        if( System.getProperty(JETTY_INTERACTIVE_NAME) != null ) isInteractive = Boolean.getBoolean(JETTY_INTERACTIVE_NAME);

        tempDir = System.getProperty("jettyTempDir");

        maxIdleTime = Integer.getInteger(JETTY_MAX_IDLE_NAME,maxIdleTime);

        jettyPort = Integer.getInteger(JETTY_PORT_NAME, jettyPort);
        jettySSLPort = Integer.getInteger(JETTY_SSL_PORT_NAME,jettySSLPort);
        usingSSL = (jettySSLPort != 0);

        if( usingSSL ) {
            keyFileName = System.getProperty(JETTY_SSL_KEY_STOREFILE_NAME,keyFileName);
            keyPassword = System.getProperty(JETTY_SSL_KEY_PASSWORD_NAME,keyPassword);
        }

        if( debug ) {
            System.out.println("=================");
            System.out.println("Jetty Embed Debug");
            System.out.println("=================");
            System.out.println("Interactive : " + isInteractive );
            System.out.println("NIO : " + usingNIO );
            System.out.println("HTTP Port : "  + jettyPort );
            System.out.println("Max Idle Time : " + maxIdleTime );
            System.out.println("SSL Port : " + (usingSSL ? jettySSLPort : "(disabled)") );
            if(usingSSL) {
                System.out.println("SSL Key Password : " + keyPassword );
                System.out.println("SSL Key File Name : " + keyFileName );
            }
            System.out.println("=================");
            System.setProperty("org.eclipse.jetty.util.log.DEBUG", "true");
        }
    }



     private static Connector buildConnector() {
        Connector connector;
        if (usingNIO) {
            connector = new SelectChannelConnector();
        } else {
            connector = new SocketConnector();
        }
        connector.setPort(jettyPort);
        connector.setMaxIdleTime(maxIdleTime);
        return connector;
    }

    private static Connector buildSslConnector() {
        SslConnector sslConnector;
        if( usingNIO ) {
            sslConnector = new SslSelectChannelConnector();
        }
        else {
            sslConnector = new SslSocketConnector();
        }
        sslConnector.setPort(jettySSLPort);
        sslConnector.setMaxIdleTime(maxIdleTime);
        sslConnector.setKeyPassword(keyPassword);
        if (keyFileName != null && keyFileName.length() != 0 ) {
            sslConnector.setKeystore(keyFileName);
        }
        return sslConnector;
    }


    private static void setThreadClassLoader() {
        Thread.currentThread().setContextClassLoader(WebAppClassLoader.class.getClassLoader());
    }

    private static WebAppContext buildContext() throws IOException {
        ProtectionDomain protectionDomain = Startup.class.getProtectionDomain();
        URL location = protectionDomain.getCodeSource().getLocation();

        WebAppContext context = new WebAppContext();
        WebAppClassLoader webAppClassLoader = new WebAppClassLoader(Startup.class.getClassLoader(),context);
        context.setClassLoader(webAppClassLoader);
        context.setContextPath(URIUtil.SLASH);
        context.setWar(location.toExternalForm());

        if( tempDir != null ) {
            File tempDirectory = new File(tempDir);
            context.setTempDirectory(tempDirectory);
        }
        return context;
    }

    public static void main(String[] args) throws Exception {

        setThreadClassLoader();
        processOptions();
        WebAppContext context = buildContext();

        if( tempDir != null ) {
            File tempDirectory = new File(tempDir);
            context.setTempDirectory(tempDirectory);
        }

        Server server = new Server();
        if( usingSSL ) {
            server.setConnectors(new Connector[]{buildConnector(),buildSslConnector()});
        }
        else {
            server.setConnectors(new Connector[]{buildConnector()});
        }
        server.setHandler(context);
        server.setSendServerVersion(false);

        run(server);

    }


    private static void run(Server server) {
        try {
            server.start();
            if( isInteractive ) {
                System.out.println("Press any key to exit.");
                System.in.read();
                server.stop();
            }
            else {
                server.join();
            }
        }
        catch (Exception e) {
            e.printStackTrace();
            System.exit(1);
        }
    }
}
